lib/std/Thread.zig

```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
     }
 }
 
+const maxNameLen = switch (std.Target.current.os.tag) {
+    .linux => 16,
+    .windows => 32,
+    .macos, .ios, .watchos, .tvos => 64,
+    .netbsd => 32,
+    .freebsd => 16,
+    .openbsd => 32,
+    else => 0,
+};
+
+pub fn setName(self: Thread, name: [:0]const u8) !void {
+    if (name.len >= maxNameLen) return error.NameTooLong;
+
+    switch (std.Target.current.os.tag) {
+        .linux => if (use_pthreads) {
+            const err = c.pthread_setname_np(self.data.handle, name.ptr);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        } else if (self.data.handle == getCurrentId()) {
+            _ = try os.prctl(.SET_NAME, .{@ptrToInt(name.ptr)});
+        } else if (!use_pthreads) {
+            var buf: [32]u8 = undefined;
+            const path = try std.fmt.bufPrint(&buf, "/proc/self/task/{d}/comm", .{self.data.handle});
+
+            const file = try std.fs.cwd().openFile(path, .{ .write = true });
+            defer file.close();
+
+            try file.writer().print("{s}", .{name});
+        },
+        .windows => {
+            // TODO
+        },
+        .macos, .ios, .watchos, .tvos => if (use_pthreads) {
+            // There doesn't seem to be a way to set the name for an arbitrary thread, only the current one.
+            if (self.data.handle != getCurrentId()) return;
+
+            const err = c.pthread_setname_np(name.ptr);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        },
+        .netbsd => if (use_pthreads) {
+            const err = c.pthread_setname_np(self.data.handle, name.ptr, null);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        },
+        .freebsd, .openbsd => if (use_pthreads) {
+            // Use pthread_set_name_np for FreeBSD because the alias pthread_setname_np is FreeBSD 12.2+ only.
+            const err = c.pthread_set_name_np(self.data.handle, name.ptr);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        },
+        else => {},
+    }
+}
+
+pub fn getName(self: Thread, allocator: *std.mem.Allocator) !?[:0]const u8 {
+    // musl doesn't provide pthread_getname_np and there's no way to retrieve the thread id of an arbitrary thread.
+    if (use_pthreads and self.data.handle != getCurrentId() and comptime std.Target.current.abi.isMusl()) {
```

Comment:
Nope, let this case use the `prctl`/`proc` code path.

It seems that [musl](https://www.openwall.com/lists/musl/2019/07/17/3) needs a bit of prodding on this front.

Reply:
I'm not sure I understand. Without this check, it would execute the `proc` reading code but I don't have access to the thread id of an arbitrary `pthread_t` so that won't work.

Reply:
Welp, I forgot the `pthread` structure type is opaque. You'd need to dig into the private definition in `pthread_impl.h` but that's pretty ugly... let's return an error instead (`Unsupported` as done for the darwin case maybe?)

Reply:
Ok, I'll change it to return an error.